### PR TITLE
Autosave

### PR DESCRIPTION
**Pull request by ntsummers1**
## Description

Is it, finally, truly, actually, happening? 

## Code author checklist
- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://bit.ly/3zPrxuZ) tests
- [ ] I have added analytics, if necessary
- [ ] I have updated the documentation, if necessary

## Reviewer checklist (two different people)
- [ ] I have done the deep review and verified the items in the above checklist are g2g
- [ ] I have done the lgtm review

### CODE REVIEW

File: services/ui-src/src/components/fields/DropdownField.tsx [hunk 1]
```diff
@@ -21,10 +22,13 @@ export const DropdownField = ({
   options,
   hint,
   nested,
+  autosave, // eslint-disable-line @typescript-eslint/no-unused-vars
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/DropdownField.tsx:25]: eslint could be removed now since we are using the var on line 91

File: services/ui-src/src/components/fields/NumberField.test.tsx [hunk 2]
```diff
@@ -197,6 +256,79 @@ describe("Test NumberField hydration functionality", () => {
...
+describe("Test NumberField component autosaves", () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+  test("NumberField autosaves with typed value when stateuser, autosave true, and form is valid", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockTrigger.mockReturnValue(true);
+    mockGetValues(undefined);
+    render(numberFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-number-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/NumberField.test.tsx:268]: nit: getByTestId

File: services/ui-src/src/components/fields/NumberField.test.tsx [hunk 3]
```diff
@@ -197,6 +256,79 @@ describe("Test NumberField hydration functionality", () => {
...
+      }
+    );
+  });
+
+  test("NumberField autosaves with default value when stateuser, autosave true, and form invalid", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockTrigger.mockReturnValue(false);
+    mockGetValues(undefined);
+    render(numberFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-number-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/NumberField.test.tsx:291]: nit: getByTestId

File: services/ui-src/src/components/fields/NumberField.test.tsx [hunk 4]
```diff
@@ -197,6 +256,79 @@ describe("Test NumberField hydration functionality", () => {
...
+        fieldData: { testNumberField: "" },
+      }
+    );
+  });
+
+  test("NumberField does not autosave when not stateuser", async () => {
+    mockedUseUser.mockReturnValue(mockAdminUser);
+    mockGetValues(undefined);
+    render(numberFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-number-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/NumberField.test.tsx:313]: nit: getByTestId

File: services/ui-src/src/components/fields/NumberField.test.tsx [hunk 5]
```diff
@@ -197,6 +256,79 @@ describe("Test NumberField hydration functionality", () => {
...
+    await userEvent.type(textField, "test value");
+    await userEvent.tab();
+    expect(mockReportContext.updateReport).toHaveBeenCalledTimes(0);
+  });
+
+  test("NumberField does not autosave if autosave is false", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockGetValues(undefined);
+    render(numberFieldComponent);
+    const textField = screen.getByTestId("test-number-field");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/NumberField.test.tsx:324]: nit: getByTestId

File: services/ui-src/src/components/fields/TextField.test.tsx [hunk 6]
```diff
@@ -79,10 +116,85 @@ describe("Test TextField hydration functionality", () => {
...
+describe("Test TextField component autosaves", () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+  test("TextField autosaves with typed value when stateuser, autosave true, and form is valid", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockTrigger.mockReturnValue(true);
+    mockGetValues(undefined);
+    render(textFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-text-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/TextField.test.tsx:128]: nit: getByTestId

File: services/ui-src/src/components/fields/TextField.test.tsx [hunk 7]
```diff
@@ -79,10 +116,85 @@ describe("Test TextField hydration functionality", () => {
...
+      }
+    );
+  });
+
+  test("TextField autosaves with default value when stateuser, autosave true, and form invalid", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockTrigger.mockReturnValue(false);
+    mockGetValues(undefined);
+    render(textFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-text-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/TextField.test.tsx:151]: nit: getByTestId

File: services/ui-src/src/components/fields/TextField.test.tsx [hunk 8]
```diff
@@ -79,10 +116,85 @@ describe("Test TextField hydration functionality", () => {
...
+        fieldData: { testTextField: "" },
+      }
+    );
+  });
+
+  test("TextField does not autosave when not stateuser", async () => {
+    mockedUseUser.mockReturnValue(mockAdminUser);
+    mockGetValues(undefined);
+    render(textFieldAutosavingComponent);
+    const textField = screen.getByTestId("test-text-field-autosave");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/TextField.test.tsx:173]: nit: getByTestId

File: services/ui-src/src/components/fields/TextField.test.tsx [hunk 9]
```diff
@@ -79,10 +116,85 @@ describe("Test TextField hydration functionality", () => {
...
+    await userEvent.type(textField, "test value");
+    await userEvent.tab();
+    expect(mockReportContext.updateReport).toHaveBeenCalledTimes(0);
+  });
+
+  test("TextField does not autosave when not autosave not set to true", async () => {
+    mockedUseUser.mockReturnValue(mockStateUser);
+    mockGetValues(undefined);
+    render(textFieldComponent);
+    const textField = screen.getByTestId("test-text-field");
```
Comment (tblackwe) [inline services/ui-src/src/components/fields/TextField.test.tsx:184]: nit: getByTestId

File: services/ui-src/src/utils/other/time.ts [hunk 10]
```diff
@@ -22,6 +22,21 @@ export const calculateTimeByType = (timeType?: string): TimeShape => {
   return timeMap?.[timeType as keyof typeof timeMap] || noon;
 };
 
+/*
+ * Returns local time in HH:mm format with the "am/pm" indicator
+ * ex: 12:02pm
+ */
+export const getLocalHourMinuteTime = () => {
```
Comment (tblackwe) [inline services/ui-src/src/utils/other/time.ts:29]: is this value presented to the user? I've been trying to figure out how the function is used.

**Review by tblackwe — APPROVED** [review]

File: services/ui-src/src/utils/other/time.ts (hunk 10 above)
Comment (tblackwe) [reply services/ui-src/src/utils/other/time.ts:29]: helps if I test locally, not on main. I figured it out.

File: services/ui-src/src/components/fields/DynamicField.tsx [hunk 11]
```diff
@@ -59,6 +62,37 @@ export const DynamicField = ({ name, label, ...props }: Props) => {
...
+      // check field data validity
+      const fieldDataIsValid = await form.trigger(event.target.name);
+      let data = displayValues;
+      // if valid, use; if not, reset to default
+      if (!fieldDataIsValid) {
+        data = displayValues.map((displayValue) => {
+          if (displayValue.id === event.target.id) displayValue.name = "";
+          return displayValue;
+        });
+      }
```
Comment (braxex) [inline services/ui-src/src/components/fields/DynamicField.tsx:77]: Comment: Can probably improve readability here and avoid mapping through all items. The effect of this is really changing `data` from `displayValues` to `displayValues` with the blurred input value changed to `""`. So we could do something like I've included below. That said, we also don't need to do this right now because the only way for a dynamic field to be invalid is if the value is blank. So we're really just changing the value from `""` to `""`. Whether existing or modified, the `if` block is only necessary if we change the dynamic field functionality in the future, so I think we can just remove it for now, along with the field validity check and instead leave a note about why it's unnecessary.

```suggestion
      // check field data validity
      const fieldDataIsValid = await form.trigger(event.target.name);
      // prepare to use display values as field value for database
      let fieldInputValues = displayValues;
      // if field not valid, reset last blurred input to default value
      if (!fieldDataIsValid) {
        const blurredInputIndex = fieldInputValues.findIndex(
          (input: EntityShape) => input.id === event.target.id
        );
        fieldInputValues[blurredInputIndex].name = "";
      }
```

File: services/ui-src/src/components/reports/StandardReportPage.test.tsx [hunk 12]
```diff
@@ -1,82 +1,34 @@
...
   RouterWrappedComponent,
 } from "utils/testing/setupJest";
 
-const mockUseNavigate = jest.fn();
-jest.mock("react-router-dom", () => ({
-  useNavigate: () => mockUseNavigate,
-  useLocation: jest.fn(() => ({
-    pathname: "/mock",
-  })),
-}));
```
Comment (braxex) [inline services/ui-src/src/components/reports/StandardReportPage.test.tsx:23]: Request: Let's keep these and add a test so we have coverage for StandardReportPage line 29 (the next navigation on the successful form check and subsequent "submission", which is just the navigation method.

File: services/ui-src/src/components/reports/StandardReportPage.test.tsx [hunk 13]
```diff
@@ -1,82 +1,34 @@
...
-    const result = render(standardPageSectionComponent);
-    const form = result.container;
-    const mockField = form.querySelector("[name='mock-text-field']")!;
-    await userEvent.type(mockField, "mock input");
-    const submitButton = form.querySelector("[type='submit']")!;
-    expect(submitButton).toBeVisible();
-    await userEvent.click(submitButton);
-    expect(mockReportContext.updateReport).toHaveBeenCalledTimes(1);
-    expect(mockUseNavigate).toHaveBeenCalledTimes(1);
-  });
```
Comment (braxex) [inline services/ui-src/src/components/reports/StandardReportPage.test.tsx:61]: Request: Related to the above comment. We can reuse this test for the navigation test. Just strip out anything related to user type and updating the fields or checking that report context was updated and we should be good to go.

File: services/ui-src/src/components/reports/StandardReportPage.tsx [hunk 14]
```diff
@@ -9,51 +9,28 @@ import {
...
-      const reportKeys = {
-        state: state,
-        id: report?.id,
-      };
-      const filteredFormData = filterFormData(enteredData, route.form.fields);
-      const dataToWrite = {
-        metadata: {
-          status: ReportStatus.IN_PROGRESS,
-          lastAlteredBy: full_name,
-        },
```
Comment (braxex) [inline services/ui-src/src/components/reports/StandardReportPage.tsx:38]: Blocking: The new autosave update methods in each field have a problem where they aren't saving the metadata because they aren't wrapped in a `metadata: {}` object like this, so `status` and `lastAlteredBy` are getting ignored by the API handler and never sent to the database.

File: services/ui-src/src/components/reports/StandardReportPage.tsx [hunk 15]
```diff
@@ -9,51 +9,28 @@ import {
...
   return (
     <Box data-testid="standard-page">
       {route.verbiage.intro && <ReportPageIntro text={route.verbiage.intro} />}
       <Form
         id={route.form.id}
         formJson={route.form}
-        onSubmit={onSubmit}
+        onSubmit={() => navigate(nextRoute)}
         formData={report?.fieldData}
+        autosave
```
Comment (braxex) [inline services/ui-src/src/components/reports/StandardReportPage.tsx:31]: Comment: Crazy how this one little line really kicks off the whole thing...

File: services/ui-src/src/utils/validation/schemas.test.ts [hunk 16]
```diff
@@ -38,7 +37,7 @@ describe("Schemas", () => {
   ];
 
   const testNumberSchema = (
-    schemaToUse: MixedSchema<any, AnyObject, any>,
+    schemaToUse: MixedSchema,
```
Comment (braxex) [inline services/ui-src/src/utils/validation/schemas.test.ts:40]: Question: I'm curious why we made this change since it's the only change to our frontend validation schema. Was it meant to be made to our backend validation schema? If so, do we need to make the same change to our frontend schema?

**Review by braxex — CHANGES_REQUESTED** [review]
Hey there! This PR is looking real good, but there are a few bugs that we need to address before we move forward. Take a look and let me know if you have any questions.

**Comment by braxex** [comment]
Note: One larger note about the way we have things constructed: I think we should extract the autosave update call logic to a utility method that would be shared by all fields. This would allow us to separate some of the duties as well. 

The field would just check if it’s an autosave field and validate that the data to be saved is valid*. Then it would call the autosave utility method. The utility method would handle checking for user type (no need for each field to pull in useUser), formatting the API handler call (no need for each field to pull in report context and statuses), and actually call the API handler (no need for each field to pull in the handler directly). *I could actually see a separate utility method checking if the data is valid and if not, setting the field to an appropriate default value. I’m on the fence about that detail though.

Since the `dataToWrite` argument of the `updateReport` API handler handles the actual merging of the data with the existing database data, our autosave utility method could accept just one prop: an object to merge into fieldData. 

This would also greatly simplify our testing files (much fewer mocks needed) and reduce the amount of duplicate code we’re adding.

**Comment by braxex** [comment]
Question: @davidkoger Currently when autosave is triggered, if field fill is valid, it is saved, and if field fill is invalid, the default initial blank state value for that field type is saved. This should work pretty well, but there's a possibility that users would overwrite existing data on accident. This is unlikely to happen for fields where it would matter more, like text area fields, and impossible to happen for fields like dropdown, radio, checkbox. 

For fields where it is more likely that a user would trigger autosave while field is in an invalid fill state (like only applicable for the following: number, date, url, email), is it acceptable that triggering autosave while invalid would result in blank value being stored?

File: services/ui-src/src/components/reports/StandardReportPage.tsx (hunk 14 above)
Comment (braxex) [reply services/ui-src/src/components/reports/StandardReportPage.tsx:38]: Update: I fixed all these in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/DynamicField.tsx (hunk 11 above)
Comment (braxex) [reply services/ui-src/src/components/fields/DynamicField.tsx]: Update: I removed this `if` block and added a comment in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/DateField.tsx [hunk 17]
```diff
@@ -46,9 +57,30 @@ export const DateField = ({
...
-  const onBlurHandler = (event: InputChangeEvent) => {
-    const fieldValue = event.target.value;
-    form.setValue(name, fieldValue, { shouldValidate: true });
+  const onBlurHandler = async (event: InputChangeEvent) => {
+    if (autosave) {
+      const { name, value } = event.target;
+      if (userIsStateUser || userIsStateRep) {
+        // check field data validity
+        const fieldDataIsValid = await form.trigger(name);
+        // if valid, use; if not, reset to default
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/DateField.tsx:67]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/app-api/utils/validation/schemaMap.ts [hunk 18]
```diff
@@ -93,18 +89,15 @@ export const ratio = () =>
...
-export const urlOptional = () => url().notRequired();
+export const urlOptional = () => url();
 
 // DATE
-export const date = () =>
-  string()
-    .required(error.REQUIRED_GENERIC)
-    .matches(dateFormatRegex, error.INVALID_DATE);
-export const dateOptional = () => date().notRequired();
+export const date = () => string().matches(dateFormatRegex, error.INVALID_DATE);
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/app-api/utils/validation/schemaMap.ts]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/DropdownField.tsx [hunk 19]
```diff
@@ -82,9 +86,36 @@ export const DropdownField = ({
...
+  const onBlurHandler = async (event: InputChangeEvent) => {
+    if (autosave) {
+      const selectedOption = {
+        label: event.target.id,
+        value: event.target.value,
+      };
+      if (userIsStateUser || userIsStateRep) {
+        // check field data validity
+        const fieldDataIsValid = await form.trigger(label);
+        // if valid, use; if not, reset to default
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/DropdownField.tsx:100]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/DateField.tsx [hunk 20]
```diff
@@ -46,9 +57,30 @@ export const DateField = ({
...
+            lastAlteredBy: full_name,
+            fieldData: { [name]: value },
+          };
+          await updateReport(reportKeys, dataToWrite);
+        }
+      }
+    } else {
+      const fieldValue = event.target.value;
+      form.setValue(name, fieldValue, { shouldValidate: true });
+    }
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/DateField.tsx]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/ChoiceListField.tsx [hunk 21]
```diff
@@ -56,6 +61,19 @@ export const ChoiceListField = ({
     }
   }, [displayValue]);
 
+  const handleUpdate = async (clickedOption: any) => {
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/ChoiceListField.tsx]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/CheckboxField.tsx [hunk 22]
```diff
@@ -18,6 +18,7 @@ export const CheckboxField = ({
         name={name}
         label={label}
         choices={choices}
+        autosave
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/CheckboxField.tsx:21]: Update: I removed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/app-api/utils/validation/schemaMap.ts [hunk 23]
```diff
@@ -36,20 +35,18 @@ const valueCleaningNumberSchema = (value: string, charsToReplace: RegExp) => {
...
+        const isValidStringValue = validNAValues.includes(value);
+        const isValidNumberValue = validNumberRegex.test(value);
+        return isValidStringValue || isValidNumberValue;
+      } else return true;
+    },
+  });
+export const numberOptional = () => number();
 
 // Number - Ratio
 export const ratio = () =>
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/app-api/utils/validation/schemaMap.ts:49]: Update: I fixed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/7bb0cbc4e95781c3724b99d5157f54c66166121a

File: services/ui-src/src/components/fields/CheckboxField.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/CheckboxField.tsx]: Update: I removed this in my commit here: https://github.com/CMSgov/mdct-mcr/pull/5336/commits/337c90399e1c62310fc7c81fd67e0341c0378ec8

**Comment by braxex** [comment]
@scwambach @ntsummers1 There are some additional areas where we can improve test coverage as well that will help us out. 

- `app-api/utils/validation/schemaMap.ts` lines 97, 106-109 might be a quick win
- `StandardReportPage` line 29 (noted above in comments as well)
- `ChoiceListField` lines 103-124 (related to the nested child field bug, so best to wait until after that's resolved)
- `Header` lines 72, 89 are related to this PR and easy to test

File: services/ui-src/src/components/fields/ChoiceListField.tsx [hunk 24]
```diff
@@ -1,16 +1,18 @@
-import { useEffect, useState } from "react";
+import { useContext, useEffect, useState } from "react";
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/ChoiceListField.tsx:1]: @scwambach @tblackwe @ntsummers1

**Review by braxex — CHANGES_REQUESTED** [review]

File: services/app-api/utils/validation/schemaMap.ts [hunk 25]
```diff
@@ -36,112 +32,114 @@ const valueCleaningNumberSchema = (value: string, charsToReplace: RegExp) => {
...
-          const isValidNumberValue = validNumberRegex.test(value);
-          return isValidStringValue || isValidNumberValue;
-        } else return true;
-      },
-    });
-export const numberOptional = () => number().notRequired();
+  string().test({
+    message: error.INVALID_NUMBER_OR_NA,
+    test: (value) => {
+      const validNumberRegex = /[0-9,.]/;
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/app-api/utils/validation/schemaMap.ts:38]: Good question. We need to allow periods and commas, but we shouldn't be allowing _only_ periods or commas. Should update the regex.

File: services/ui-src/src/components/fields/ChoiceListField.test.tsx [hunk 26]
```diff
@@ -88,173 +104,581 @@ const mockChoiceWithChild = {
...
 describe("Test ChoiceListField component rendering", () => {
+  it("ChoiceList should render a normal Radiofield that doesn't have children", () => {
+    render(RadioComponent);
+    expect(screen.getByText("Choice 1")).toBeVisible();
+    expect(screen.getByText("Choice 2")).toBeVisible();
+  });
+
+  it("ChoiceList should render a normal Checkbox that doesn't have children", () => {
+    render(CheckboxComponent);
+    expect(screen.getByText("Choice 1")).toBeVisible();
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/ChoiceListField.test.tsx:159]: Agreed. This would be a good update across these test files.

File: services/ui-src/src/components/fields/ChoiceListField.test.tsx [hunk 27]
```diff
@@ -88,173 +104,581 @@ const mockChoiceWithChild = {
...
+    value: "Other, specify",
+    name: "checkboxWithTextChild",
+    checked: false,
+  };
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe("Checkbox Choice that has no children", () => {
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/ChoiceListField.test.tsx:560]: agreed. let's do it.

File: services/ui-src/src/components/fields/ChoiceListField.tsx [hunk 28]
```diff
@@ -120,19 +126,51 @@ export const ChoiceListField = ({
...
+        ];
+        const reportArgs = { id: report?.id, updateReport };
+        const user = { userName: full_name, state };
+        await autosaveFieldData({
+          form,
+          fields,
+          report: reportArgs,
+          user,
+        });
+      }, 200);
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/ChoiceListField.tsx:173]: agreed.

File: services/ui-src/src/components/fields/NumberField.test.tsx [hunk 29]
```diff
@@ -89,7 +125,11 @@ describe("Test Comma-Separated Masked NumberField", () => {
 });
 
 describe("Test Currency Masked NumberField", () => {
+  afterEach(() => {
```
[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/NumberField.test.tsx:128]: agreed. let's do it.

File: services/ui-src/src/utils/validation/schemas.test.ts (hunk 16 above)
Comment (braxex) [reply services/ui-src/src/utils/validation/schemas.test.ts:40]: yeah this is fine.

File: services/ui-src/src/components/reports/StandardReportPage.test.tsx (hunk 12 above)
Comment (braxex) [reply services/ui-src/src/components/reports/StandardReportPage.test.tsx:23]: yeah we're fine here.

File: services/ui-src/src/components/reports/StandardReportPage.test.tsx (hunk 13 above)
Comment (braxex) [reply services/ui-src/src/components/reports/StandardReportPage.test.tsx:61]: yeah this is fine.

**Review by scwambach — APPROVED** [review]
looked through all the code, pulled it down, ran it, been apart of the whole journey, I very much approve of this

**Review by tblackwe — APPROVED** [review]

**Review by braxex — APPROVED** [review]
